File: store/src/test/java/com/nytimes/android/external/store3/StoreTest.java
package com.nytimes.android.external.store3;

import com.nytimes.android.external.cache3.Cache;
import com.nytimes.android.external.cache3.CacheBuilder;
import com.nytimes.android.external.store.util.Result;
import com.nytimes.android.external.store3.base.Fetcher;
import com.nytimes.android.external.store3.base.Persister;
import com.nytimes.android.external.store3.base.impl.BarCode;
import com.nytimes.android.external.store3.base.impl.RealStore;
import com.nytimes.android.external.store3.base.impl.Store;
import com.nytimes.android.external.store3.base.impl.StoreBuilder;
import com.nytimes.android.external.store3.util.NoopPersister;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import io.reactivex.Maybe;
import io.reactivex.Single;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class StoreTest {

    private static final String DISK = "disk";
    private static final String NETWORK = "fresh";
    private static final String MEMORY = "memory";
    final AtomicInteger counter = new AtomicInteger(0);
    @Mock
    Fetcher<String, BarCode> fetcher;
    @Mock
    Persister<String, BarCode> persister;
    private final BarCode barCode = new BarCode("key", "value");

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    public void testSimple() {

        Store<String, BarCode> simpleStore = StoreBuilder.<String>barcode()
                .persister(persister)
                .fetcher(fetcher)
                .open();


        when(fetcher.fetch(barCode))
                .thenReturn(Single.just(NETWORK));

        when(persister.read(barCode))
                .thenReturn(Maybe.<String>empty())
                .thenReturn(Maybe.just(DISK));

        when(persister.write(barCode, NETWORK))
                .thenReturn(Single.just(true));

        String value = simpleStore.get(barCode).blockingGet();

        assertThat(value).isEqualTo(DISK);
        value = simpleStore.get(barCode).blockingGet();
        assertThat(value).isEqualTo(DISK);
        verify(fetcher, times(1)).fetch(barCode);
    }

    @Test
    public void testSimpleWithResult() {

        Store<String, BarCode> simpleStore = StoreBuilder.<String>barcode()
            .persister(persister)
            .fetcher(fetcher)
            .open();


        when(fetcher.fetch(barCode))
            .thenReturn(Single.just(NETWORK));

        when(persister.read(barCode))
            .thenReturn(Maybe.<String>empty())
            .thenReturn(Maybe.just(DISK));

        when(persister.write(barCode, NETWORK))
            .thenReturn(Single.just(true));

        Result<String> result = simpleStore.getWithResult(barCode).blockingGet();

        assertThat(result.source()).isEqualTo(Result.Source.NETWORK);
        assertThat(result.value()).isEqualTo(DISK);

        result = simpleStore.getWithResult(barCode).blockingGet();
        assertThat(result.source()).isEqualTo(Result.Source.CACHE);
        assertThat(result.value()).isEqualTo(DISK);
        verify(fetcher, times(1)).fetch(barCode);
    }


    @Test
    public void testDoubleTap() {

        Store<String, BarCode> simpleStore = StoreBuilder.<String>barcode()
                .persister(persister)
                .fetcher(fetcher)
                .open();

        Single<String> networkSingle =
                Single.create(emitter -> {
                    if (counter.incrementAndGet() == 1) {
                        emitter.onSuccess(NETWORK);
                    } else {
                        emitter.onError(new RuntimeException("Yo Dawg your inflight is broken"));
                    }
                });


        when(fetcher.fetch(barCode))
                .thenReturn(networkSingle);

        when(persister.read(barCode))
                .thenReturn(Maybe.<String>empty())
                .thenReturn(Maybe.just(DISK));

        when(persister.write(barCode, NETWORK))
                .thenReturn(Single.just(true));


        String response = simpleStore.get(barCode)
                .zipWith(simpleStore.get(barCode), (s, s2) -> "hello")
                .blockingGet();
        assertThat(response).isEqualTo("hello");
        verify(fetcher, times(1)).fetch(barCode);
    }

    @Test
    public void testDoubleTapWithResult() {

        Store<String, BarCode> simpleStore = StoreBuilder.<String>barcode()
            .persister(persister)
            .fetcher(fetcher)
            .open();

        Single<String> networkSingle =
            Single.create(emitter -> {
                if (counter.incrementAndGet() == 1) {
                    emitter.onSuccess(NETWORK);
                } else {
                    emitter.onError(new RuntimeException("Yo Dawg your inflight is broken"));
                }
            });


        when(fetcher.fetch(barCode))
            .thenReturn(networkSingle);

        when(persister.read(barCode))
            .thenReturn(Maybe.<String>empty())
            .thenReturn(Maybe.just(DISK));

        when(persister.write(barCode, NETWORK))
            .thenReturn(Single.just(true));


        Result<String> response = simpleStore.getWithResult(barCode)
            .zipWith(simpleStore.getWithResult(barCode), (s, s2) -> Result.createFromNetwork("hello"))
            .blockingGet();

        assertThat(response.source()).isEqualTo(Result.Source.NETWORK);
        assertThat(response.value()).isEqualTo("hello");
        verify(fetcher, times(1)).fetch(barCode);
    }


    @Test
    public void testSubclass() {

        RealStore<String, BarCode> simpleStore = new SampleStore(fetcher, persister);
        simpleStore.clear();

        when(fetcher.fetch(barCode))
                .thenReturn(Single.just(NETWORK));

        when(persister.read(barCode))
                .thenReturn(Maybe.<String>empty())
                .thenReturn(Maybe.just(DISK));
        when(persister.write(barCode, NETWORK)).thenReturn(Single.just(true));

        String value = simpleStore.get(barCode).blockingGet();
        assertThat(value).isEqualTo(DISK);
        value = simpleStore.get(barCode).blockingGet();
        assertThat(value).isEqualTo(DISK);
        verify(fetcher, times(1)).fetch(barCode);
    }

    @Test
    public void testSubclassWithResult() {

        RealStore<String, BarCode> simpleStore = new SampleStore(fetcher, persister);
        simpleStore.clear();

        when(fetcher.fetch(barCode))
            .thenReturn(Single.just(NETWORK));

        when(persister.read(barCode))
            .thenReturn(Maybe.<String>empty())
            .thenReturn(Maybe.just(DISK));
        when(persister.write(barCode, NETWORK)).thenReturn(Single.just(true));

        Result<String> result = simpleStore.getWithResult(barCode).blockingGet();

        assertThat(result.source()).isEqualTo(Result.Source.NETWORK);
        assertThat(result.value()).isEqualTo(DISK);

        result = simpleStore.getWithResult(barCode).blockingGet();
        assertThat(result.source()).isEqualTo(Result.Source.CACHE);
        assertThat(result.value()).isEqualTo(DISK);
        verify(fetcher, times(1)).fetch(barCode);
    }

    @Test
    public void testNoopAndDefault() {

        Persister<String, BarCode> persister = spy(NoopPersister.Companion.<String, BarCode>create());
        RealStore<String, BarCode> simpleStore = new SampleStore(fetcher, persister);


        when(fetcher.fetch(barCode))
                .thenReturn(Single.just(NETWORK));

        String value = simpleStore.get(barCode).blockingGet();
        verify(fetcher, times(1)).fetch(barCode);
        verify(persister, times(1)).write(barCode, NETWORK);
        verify(persister, times(2)).read(barCode);
        assertThat(value).isEqualTo(NETWORK);


        value = simpleStore.get(barCode).blockingGet();
        verify(persister, times(2)).read(barCode);
        verify(persister, times(1)).write(barCode, NETWORK);
        verify(fetcher, times(1)).fetch(barCode);

        assertThat(value).isEqualTo(NETWORK);
    }

    @Test
    public void testNoopAndDefaultWithResult() {

        Persister<String, BarCode> persister = spy(NoopPersister.Companion.<String, BarCode>create());
        RealStore<String, BarCode> simpleStore = new SampleStore(fetcher, persister);


        when(fetcher.fetch(barCode))
            .thenReturn(Single.just(NETWORK));

        Result<String> value = simpleStore.getWithResult(barCode).blockingGet();
        verify(fetcher, times(1)).fetch(barCode);
        verify(persister, times(1)).write(barCode, NETWORK);
        verify(persister, times(2)).read(barCode);
        assertThat(value.source()).isEqualTo(Result.Source.NETWORK);
        assertThat(value.value()).isEqualTo(NETWORK);


        value = simpleStore.getWithResult(barCode).blockingGet();
        verify(persister, times(2)).read(barCode);
        verify(persister, times(1)).write(barCode, NETWORK);
        verify(fetcher, times(1)).fetch(barCode);

        assertThat(value.source()).isEqualTo(Result.Source.CACHE);
        assertThat(value.value()).isEqualTo(NETWORK);
    }

    @Test
    public void testEquivalence() {
        Cache<BarCode, String> cache = CacheBuilder.newBuilder()
                .maximumSize(1)
                .expireAfterAccess(Long.MAX_VALUE, TimeUnit.SECONDS)
                .build();

        cache.put(barCode, MEMORY);
        String value = cache.getIfPresent(barCode);
        assertThat(value).isEqualTo(MEMORY);

        value = cache.getIfPresent(new BarCode(barCode.getType(), barCode.getKey()));
        assertThat(value).isEqualTo(MEMORY);
    }
}

File: store/src/test/java/com/nytimes/android/external/store3/StoreTest.kt
package com.nytimes.android.external.store3

import com.nhaarman.mockitokotlin2.*
import com.nytimes.android.external.cache3.CacheBuilder
import com.nytimes.android.external.store3.base.Fetcher
import com.nytimes.android.external.store3.base.Persister
import com.nytimes.android.external.store3.base.impl.BarCode
import com.nytimes.android.external.store3.base.impl.StoreBuilder
import com.nytimes.android.external.store3.util.NoopPersister
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.Test
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

class StoreTest {
    val counter = AtomicInteger(0)
    val fetcher: Fetcher<String, BarCode> = mock()
    var persister: Persister<String, BarCode> = mock()
    private val barCode = BarCode("key", "value")

    @Test
    fun testSimple() = runBlocking<Unit> {

        val simpleStore = StoreBuilder.barcode<String>()
                .persister(persister)
                .fetcher(fetcher)
                .open()


        whenever(fetcher.fetch(barCode))
                .thenReturn(NETWORK)

        whenever(persister.read(barCode))
                .thenReturn(null)
                .thenReturn(DISK)

        whenever(persister.write(barCode, NETWORK))
                .thenReturn(true)

        var value = simpleStore.get(barCode)

        assertThat(value).isEqualTo(DISK)
        value = simpleStore.get(barCode)
        assertThat(value).isEqualTo(DISK)
        verify(fetcher, times(1)).fetch(barCode)
    }

    @Test
    fun testDoubleTap() = runBlocking<Unit> {

        val simpleStore = StoreBuilder.barcode<String>()
                .persister(persister)
                .fetcher(fetcher)
                .open()

        whenever(fetcher.fetch(barCode))
                .thenAnswer {
                    if (counter.incrementAndGet() == 1) {
                        NETWORK
                    } else {
                        throw RuntimeException("Yo Dawg your inflight is broken")
                    }
                }

        whenever(persister.read(barCode))
                .thenReturn(null)
                .thenReturn(DISK)

        whenever(persister.write(barCode, NETWORK))
                .thenReturn(true)


        val deferred = async { simpleStore.get(barCode) }
        simpleStore.get(barCode)
        deferred.await()

        verify(fetcher, times(1)).fetch(barCode)
    }

    @Test
    fun testSubclass() = runBlocking<Unit> {

        val simpleStore = SampleStore(fetcher, persister)
        simpleStore.clearMemory()

        whenever(fetcher.fetch(barCode))
                .thenReturn(NETWORK)

        whenever(persister.read(barCode))
                .thenReturn(null)
                .thenReturn(DISK)
        whenever(persister.write(barCode, NETWORK)).thenReturn(true)

        var value = simpleStore.get(barCode)
        assertThat(value).isEqualTo(DISK)
        value = simpleStore.get(barCode)
        assertThat(value).isEqualTo(DISK)
        verify(fetcher, times(1)).fetch(barCode)
    }

    @Test
    fun testNoopAndDefault() = runBlocking<Unit> {

        val persister = spy(NoopPersister.create<String, BarCode>())
        val simpleStore = SampleStore(fetcher, persister)


        whenever(fetcher.fetch(barCode))
                .thenReturn(NETWORK)

        var value = simpleStore.get(barCode)
        verify(fetcher, times(1)).fetch(barCode)
        verify(persister, times(1)).write(barCode, NETWORK)
        verify(persister, times(2)).read(barCode)
        assertThat(value).isEqualTo(NETWORK)


        value = simpleStore.get(barCode)
        verify(persister, times(2)).read(barCode)
        verify(persister, times(1)).write(barCode, NETWORK)
        verify(fetcher, times(1)).fetch(barCode)

        assertThat(value).isEqualTo(NETWORK)
    }

    @Test
    fun testEquivalence() = runBlocking<Unit> {
        val cache = CacheBuilder.newBuilder()
                .maximumSize(1)
                .expireAfterAccess(java.lang.Long.MAX_VALUE, TimeUnit.SECONDS)
                .build<BarCode, String>()

        cache.put(barCode, MEMORY)
        var value = cache.getIfPresent(barCode)
        assertThat(value).isEqualTo(MEMORY)

        value = cache.getIfPresent(BarCode(barCode.type, barCode.key))
        assertThat(value).isEqualTo(MEMORY)
    }

    @Test
    fun testFreshUsesOnlyNetwork() = runBlocking<Unit> {
        val simpleStore = StoreBuilder.barcode<String>()
                .persister(persister)
                .fetcher(fetcher)
                .networkBeforeStale()
                .open()

        whenever(fetcher.fetch(barCode)) doThrow RuntimeException(ERROR)

        whenever(persister.read(barCode)) doReturn DISK

        try {
            simpleStore.fresh(barCode)
            fail("Exception not thrown!")
        } catch (e: Exception) {
            assertThat(e.message).isEqualTo(ERROR)
        }

        verify(fetcher, times(1)).fetch(barCode)
        verify(persister, never()).read(any())
    }

    companion object {
        private const val DISK = "disk"
        private const val NETWORK = "fresh"
        private const val MEMORY = "memory"
        private const val ERROR = "error!!"
    }
}
